### PR TITLE
Trace ID aware load-balancing exporter - 3/4

### PR DESCRIPTION
**Pull request by jpkrohling**
**Description:** 

* Added the backend resolver
* Added the metrics definitions

**Link to tracking Issue:** Partially solves open-telemetry/opentelemetry-collector#1724, next step after #1349

**Testing:** unit tests

**Documentation:** godoc

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
@nilebox , @bogdandrutu , could one of you please review this? It's independent of the second PR (#1349). Once it's approved, I'll fix the conflicts with go.mod/sum.

**Comment by jpkrohling** [comment]
ping @nilebox , @bogdandrutu, @tigrannajaryan

**Comment by jpkrohling** [comment]
Rebased and ready to be reviewed.

**Comment by jpkrohling** [comment]
ping @nilebox , @bogdandrutu, @tigrannajaryan

**Comment by jpkrohling** [comment]
ping @nilebox , @bogdandrutu, @tigrannajaryan

File: exporter/loadbalancingexporter/resolver_dns.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+)
+
+var _ resolver = (*dnsResolver)(nil)
+
+const (
+	defaultResInterval = 5 * time.Second
+	defaultResTimeout  = time.Second
```
Comment (tigrannajaryan) [inline exporter/loadbalancingexporter/resolver_dns.go:35]: Is this reasonably enough time to resolve a hostname? I'd imagine it is quite possible to take more time.

File: exporter/loadbalancingexporter/resolver_dns.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+		return r.endpoints, nil
+	}
+
+	// the list has changed!
+	r.updateLock.Lock()
+	r.endpoints = backends
+	r.updateLock.Unlock()
+	stats.Record(mCtx, mNumBackends.M(int64(len(backends))))
+
+	// propate the change
```
Comment (tigrannajaryan) [inline exporter/loadbalancingexporter/resolver_dns.go:148]:
```suggestion
	// propagate the change
```

File: exporter/loadbalancingexporter/resolver_static.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	return &staticResolver{
+		endpoints: endpoints,
+	}, nil
+}
+
+func (r *staticResolver) start(ctx context.Context) error {
+	r.resolve(ctx) // right now, this can't fail
+	return nil
```
Comment (tigrannajaryan) [inline exporter/loadbalancingexporter/resolver_static.go:48]: Does this mean that is onChange is called after start the callback will never be called? Is this expected behavior?

File: exporter/loadbalancingexporter/resolver.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	// returns either a non-nil error and a nil list of endpoints, or a non-nil list of endpoints and nil error.
+	resolve(context.Context) ([]string, error)
+
+	// start signals the resolver to start its work
+	start(context.Context) error
+
+	// shutdown signals the resolver to finish its work
+	shutdown(context.Context) error
+
+	// onChange registers a function to call back whenever the list of backends is updated
```
Comment (tigrannajaryan) [inline exporter/loadbalancingexporter/resolver.go:31]: 1. Can this be called before `start`?
2. If it is called after `start` is it expected that the callback will be called immediately or soon after with the current list of backends?

File: exporter/loadbalancingexporter/resolver.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+type resolver interface {
+	// resolve returns the current list of endpoints.
+	// returns either a non-nil error and a nil list of endpoints, or a non-nil list of endpoints and nil error.
+	resolve(context.Context) ([]string, error)
+
+	// start signals the resolver to start its work
+	start(context.Context) error
+
+	// shutdown signals the resolver to finish its work
+	shutdown(context.Context) error
```
Comment (tigrannajaryan) [inline exporter/loadbalancingexporter/resolver.go:29]: Does this provide any guarantees that callbacks will not be called after this call returns or it is undefined behavior?

File: exporter/loadbalancingexporter/resolver_dns.go (hunk 1 above)
Comment (jpkrohling) [reply exporter/loadbalancingexporter/resolver_dns.go:35]: In the common case, 100ms is already considered a high latency: either the authoritative server is close enough (like in Kubernetes), or caches are expected to exist. But even if it fails once or twice in a row, it's typically not a problem, as changes aren't happening every second.

File: exporter/loadbalancingexporter/resolver_static.go (hunk 3 above)
Comment (jpkrohling) [reply exporter/loadbalancingexporter/resolver_static.go:48]: Yes, for the static resolver, this is the expected behavior. There's no need to continuously resolve the same list of static hosts.

File: exporter/loadbalancingexporter/resolver.go (hunk 4 above)
Comment (jpkrohling) [reply exporter/loadbalancingexporter/resolver.go:31]: I added a comment making it clear that `onChange` has to be called before `start`, otherwise the first update might be missed.

So:

1. yes, in fact, it *should* be called before start (not "must", but "should")
1. no, this is only called when the resolver gets a different set of backends

File: exporter/loadbalancingexporter/resolver.go (hunk 5 above)
Comment (jpkrohling) [reply exporter/loadbalancingexporter/resolver.go:29]: Great question, I added a comment here with the expected behavior, but in short: yes. It's currently only relevant for the DNS resolver, and here's what it does:

1. while a resolution is happening, a wait group is incremented
1. once the resolution finishes, the wait group is decreased
1. if a shutdown happens, a message is sent to the stop channel, which is in the same select as the resolve function, so, it will not happen concurrently to the steps above
1. once the periodic resolution gets the stop message, it returns
1. the shutdown function will send a message to the channel and wait for the wait group to be completed

In short: a client might call shutdown, and it will block until the current resolution has finished. During this block, it might happen that the resolution triggers the `onChange` callbacks, but it will not trigger them once shutdown returns.

File: exporter/loadbalancingexporter/resolver_dns.go (hunk 2 above)
Comment (jpkrohling) [reply exporter/loadbalancingexporter/resolver_dns.go]: PR updated.

**Comment by jpkrohling** [comment]
I'll rebase once it passes the code review, or if there are other requested changes.

File: exporter/loadbalancingexporter/resolver_static.go (hunk 3 above)
Comment (tigrannajaryan) [reply exporter/loadbalancingexporter/resolver_static.go:48]: Please document the behavior.

File: exporter/loadbalancingexporter/resolver_dns.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	go r.periodicallyResolve()
+
+	return nil
+}
+
+func (r *dnsResolver) shutdown(ctx context.Context) error {
+	close(r.stopCh)
+	r.shutdownWg.Wait()
```
Comment (tigrannajaryan) [inline exporter/loadbalancingexporter/resolver_dns.go:87]: Let's say shutdown() is called when resolve() is about to execute(). If resolve() is on line 105, but not yet executed it, then here line 87 will not wait for anything since shutdownWg counter is 0. shutdown() will return without waiting, then resolve() will be successfully executed and will call onChange callbacks, which is a violation of the resolver interface contract.
I may be wrong in the analysis, please double check.

Comment (jpkrohling) [reply exporter/loadbalancingexporter/resolver_dns.go:87]: You are correct in the analysis. It's a fairly low probability of happening, as this is not in the critical path and two concurrent registers have to be at the exact sweet spot for this to happen. In the typical case, the side-effect also won't matter much, as the resolver will only shutdown when the exporter is shutting down, which in turn only happens when the whole collector is shutting down. This becomes a relevant problem once we have the ability to start/stop components on demand, independent on the collector's life cycle.

In any case, we have two possible solutions here:

1. clear the callbacks on shutdown (my preferred)
1. share a lock instead of wait group.

I can't find a drawback on the first solution, so, I'm adopting it as it's easier to read, understand and reason about.

**Comment by jpkrohling** [comment]
PR updated, addressing the review comments.

File: exporter/loadbalancingexporter/resolver_dns.go (hunk 6 above)
Comment (tigrannajaryan) [reply exporter/loadbalancingexporter/resolver_dns.go:88]: >as the resolver will only shutdown when the exporter is shutting down, which in turn only happens when the whole collector is shutting down.

It can also happen during hot reload on config change that we plan to add in the future so we need to support this cleanly.

>clear the callbacks on shutdown (my preferred)

Sounds good.

File: exporter/loadbalancingexporter/resolver_dns.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+		return err
+	}
+
+	go r.periodicallyResolve()
+
+	return nil
+}
+
+func (r *dnsResolver) shutdown(ctx context.Context) error {
+	r.onChangeCallbacks = nil
```
Comment (tigrannajaryan) [inline exporter/loadbalancingexporter/resolver_dns.go:86]: I think we need to protect this field, this is a data race with line 150 which read it.

**Comment by tigrannajaryan** [comment]
LGTM, except one minor comment.

File: exporter/loadbalancingexporter/resolver_dns.go (hunk 7 above)
Comment (jpkrohling) [reply exporter/loadbalancingexporter/resolver_dns.go:88]: Done

**Comment by jpkrohling** [comment]
PR updated.

**Review by tigrannajaryan — APPROVED** [review]
LGTM
